### PR TITLE
Merge and update variable and params annotations from typeshed

### PR DESCRIPTION
**Pull request by Avasam**
## Summary of changes

Merge and update (because they were not all correct or exact) variable and params annotations from typeshed. Missing public annotations are missing from typeshed too.
This PR purposefully avoids adding TypeVars, overloads and return types. Those will all be tackled in a follow-up PR. Return types are inferred at this step.

Step 3.1 of https://github.com/pypa/setuptools/issues/2345#issuecomment-1627561833 but only for `pkg_resources`

### Pull Request Checklist
- [x] Changes have tests (type-checking tests are coming right after #3979 )
- [x] News fragment added in [`newsfragments/`].
  _(See [documentation][PR docs] for details)_

[`newsfragments/`]: https://github.com/pypa/setuptools/tree/master/newsfragments
[PR docs]:
https://setuptools.pypa.io/en/latest/development/developer-guide.html#making-a-pull-request

### CODE REVIEW

File: pkg_resources/__init__.py
Comment (Avasam) [inline pkg_resources/__init__.py:137]: These aliases come from the typeshed stub. Please lmk if you'd prefer different names

**Review by jaraco — APPROVED** [review]

File: pkg_resources/__init__.py
Comment (Avasam) [inline pkg_resources/__init__.py:111]: I feel like the way it's been done in
https://github.com/pypa/setuptools/pull/4242/files#diff-e3d3d454fa3a072c9f46f8affa27513892fbc2d245e87f57a5927a7be851de05R3332-R3349
is better than casting all these with forward references and redefining callable types.

File: pkg_resources/__init__.py
Comment (Avasam) [inline pkg_resources/__init__.py:1744]:
```suggestion
```
Actually there's no assurance that an egg is even found.

**Comment by Avasam** [comment]
As draft until https://github.com/pypa/setuptools/pull/4262 is completed

File: pkg_resources/__init__.py
Comment (abravalheri) [inline pkg_resources/__init__.py:2117]: Instead of `_AdapterType`, it would be more "sound" here to have a `callable that, passed a path item and the importer instance, yields ``Distribution`` instances found on that path item` or something that matches the same type as the `_distribution_finders` variable above, would it?

File: pkg_resources/__init__.py
Comment (abravalheri) [inline pkg_resources/__init__.py:2299]: Instead of `_AdapterType`, it would be more "sound" here to have a callable like this:

```python
        def namespace_handler(importer, path_entry, moduleName, module):
            # return a path_entry to use for child packages
```

wouldn't it?

File: pkg_resources/__init__.py
Comment (abravalheri) [inline pkg_resources/__init__.py:3336]: Here the way to go is probably have 3 definitions using `typing.overload` for the 3 different types of adapters and compatible `ob` (something like `_ProviderFactory`+ `_LoaderProtocol`; `_DistributionFider`+`_Importer`; `_NamespaceHandler`+`_Importer`, with `_Importer` a protocol that is compatible with `PathEntryFinder` and/or `MetaPathFinder`? ... considering the limitations in https://github.com/python/typeshed/issues/11541 and https://github.com/python/typeshed/issues/2468)

File: pkg_resources/__init__.py
Comment (Avasam) [reply pkg_resources/__init__.py:139]: I plan on updating with:
```py
_ProviderFactoryType = Callable[[_ModuleLike], "IResourceProvider"]
_DistFinderType = Callable[[_T, str, bool], Iterable["Distribution"]]
_NSHandlerType = Callable[[_T, str, str, types.ModuleType], Optional[str]]
_AdapterT = TypeVar(
    "_AdapterT", _DistFinderType[Any], _ProviderFactoryType, _NSHandlerType[Any]
)
```

in a follow-up PR, but I can add it to this one if you'd like.

File: pkg_resources/__init__.py
Comment (Avasam) [reply pkg_resources/__init__.py:393]: See comment https://github.com/pypa/setuptools/pull/4246/files#r1596038268

File: pkg_resources/__init__.py
Comment (Avasam) [reply pkg_resources/__init__.py:128]: That comes from typeshed. It's private to help prevent accidental runtime imports because typeshed didn't start using `@type_check_only` until very recently (and I believe we'll keep these names private anyway). In other words, `_LoaderProtocol` doesn't exist in the `types` module.

We could potentially ask to move `_LoaderProtocol` from `types.pyi` to `typeshed/__init__.pyi` and see what other typeshed maintainers think.

File: pkg_resources/__init__.py
Comment (Avasam) [reply pkg_resources/__init__.py:404]: See comment https://github.com/pypa/setuptools/pull/4246/files#r1596038268
This would become
```py
def register_loader_type(
    loader_type: Type[_ModuleLike],
    provider_factory: _ProviderFactoryType,
) -> None: ...
```

File: pkg_resources/__init__.py
Comment (Avasam) [reply pkg_resources/__init__.py:2117]: See comment https://github.com/pypa/setuptools/pull/4246/files#r1596038268
This would become:
```py
def register_finder(
    importer_type: Type[_T], distribution_finder: _DistFinderType[_T]
) -> None:
```

File: pkg_resources/__init__.py
Comment (Avasam) [reply pkg_resources/__init__.py:2299]: See comment https://github.com/pypa/setuptools/pull/4246/files#r1596038268
This would become:
```py
def register_namespace_handler(
    importer_type: Type[_T],
    namespace_handler: _NSHandlerType[_T],
) -> None: ...
```

File: pkg_resources/extern/__init__.py
Comment (Avasam) [reply pkg_resources/extern/__init__.py:73]:
```suggestion
            # https://github.com/python/typeshed/issues/11882
```
Whatever the result now linking to this issue and its resolution should be correct.

File: pkg_resources/__init__.py
Comment (abravalheri) [reply pkg_resources/__init__.py:128]: If `_LoaderProtocol` is undocumented, I think we should not import it.

If `typeshed` maintainers are happy to make an utility definition of `LoaderProtocol` in [`_typeshed`](https://github.com/python/typeshed/tree/main/stdlib/_typeshed) accessible for public usage that would be great. Otherwise if the protocol is small, declaring the protocol inline should also be fine.

(Brain dump: probably having an `_typeshed.importlib` module that helps to address the concerns in https://github.com/python/typeshed/issues/11882#issuecomment-2104253881 would be even greater).

File: pkg_resources/__init__.py
Comment (abravalheri) [reply pkg_resources/__init__.py:139]: If you are planning that as a follow-up PR, that should be fine.

Thank you very much.

**Comment by Avasam** [comment]
@abravalheri in-code comments have been updated to reflect that fixes have been done upstream. I don't think we need to wait for the next mypy release. `type-ignore` can be removed when updating to mypy 1.11

**Comment by abravalheri** [comment]
Thank you very much.

**Pull request merged**
